provider/server/machine_volume_detach_test.go

```diff
@@ -0,0 +1,211 @@
+		Expect(err).NotTo(HaveOccurred())
+		Expect(domainXMLData).NotTo(BeEmpty())
+
+		By("ensuring domain for machine is in running state")
+		Eventually(func() libvirt.DomainState {
+			domainState, _, err := libvirtConn.DomainGetState(domain, 0)
+			Expect(err).NotTo(HaveOccurred())
+			return libvirt.DomainState(domainState)
+		}).Should(Equal(libvirt.DomainRunning))
+
+		By("ensuring machine is in running state and other status fields have been updated")
+		Eventually(func() *iri.MachineStatus {
+			listResp, err := machineClient.ListMachines(ctx, &iri.ListMachinesRequest{
+				Filter: &iri.MachineFilter{
+					Id: createResp.Machine.Metadata.Id,
+				},
+			})
+			Expect(err).NotTo(HaveOccurred())
+			Expect(listResp.Machines).NotTo(BeEmpty())
+			Expect(listResp.Machines).Should(HaveLen(1))
+			return listResp.Machines[0].Status
+		}).Should(SatisfyAll(
+			HaveField("Volumes", ContainElements(
+				&iri.VolumeStatus{
+					Name:   "disk-1",
+					Handle: "libvirt-provider.ironcore.dev/empty-disk/disk-1",
+					State:  iri.VolumeState_VOLUME_ATTACHED,
+				},
+				&iri.VolumeStatus{
+					Name:   "disk-2",
+					Handle: "libvirt-provider.ironcore.dev/empty-disk/disk-2",
+					State:  iri.VolumeState_VOLUME_ATTACHED,
+				},
+				&iri.VolumeStatus{
+					Name:   "volume-1",
+					Handle: "libvirt-provider.ironcore.dev/ceph/libvirt-provider.ironcore.dev/ceph^dummy",
+					State:  iri.VolumeState_VOLUME_ATTACHED,
+				})),
+			HaveField("State", Equal(iri.MachineState_MACHINE_RUNNING)),
+		))
+
+		By("ensuring both the empty disks and a ceph volume is attached to a machine domain")
+		var disks []libvirtxml.DomainDisk
+		Eventually(func() int {
+			domainXMLData, err := libvirtConn.DomainGetXMLDesc(domain, 0)
+			Expect(err).NotTo(HaveOccurred())
+			domainXML := &libvirtxml.Domain{}
+			Expect(domainXML.Unmarshal(domainXMLData)).Should(Succeed())
+			disks = domainXML.Devices.Disks
+			return len(disks)
+		}).Should(Equal(4))
+		Expect(disks[0].Serial).To(HavePrefix("oda"))
+		Expect(disks[1].Serial).To(HavePrefix("odb"))
+		Expect(disks[2].Serial).To(HavePrefix("odc"))
+
+		// wait for boot image to be available before detaching volume
+		time.Sleep(20 * time.Second)
+
+		By("detaching disk-1 from machine")
+		diskDetachResp, err := machineClient.DetachVolume(ctx, &iri.DetachVolumeRequest{
+			MachineId: createResp.Machine.Metadata.Id,
+			Name:      "disk-1",
+		})
+		Expect(err).NotTo(HaveOccurred())
+		Expect(diskDetachResp).NotTo(BeNil())
+
+		By("ensuring disk-1 is unplugged from a machine domain")
+		Eventually(func() int {
+			domainXMLData, err := libvirtConn.DomainGetXMLDesc(domain, 0)
+			Expect(err).NotTo(HaveOccurred())
+			domainXML := &libvirtxml.Domain{}
+			err = domainXML.Unmarshal(domainXMLData)
+			Expect(err).NotTo(HaveOccurred())
+			disks = domainXML.Devices.Disks
+			return len(disks)
+		}).Should(Equal(3))
+
+		// wait for boot image to be available before detaching volume
+		time.Sleep(20 * time.Second)
```

Comment:
why do you have second time.Sleep here?

Reply:
After detaching disk-1, it takes time to reconcile machine, and then only ceph volume can be detached so it's added to wait to complete machine reconciliation.

Reply:
i recommend update comment in this case because it is saying something different.